### DISCUSSION TITLE
Seperate Admin UI and Admin API

### DISCUSSION THREAD

**Discussion by shavidze** [medusajs/medusa#5898] [Q&A]
Is there any **separate project,** or anything related to it to run separately `admin UI` and `admin API` projects?

P.S. There was a separate project but it's archived for now https://github.com/medusajs/admin

**Comment by shavidze** [comment]
@shahednasser please, help us

**Reply by shahednasser** [reply]
Not sure what you mean. What do you exactly want to run separately and why?

**Reply by shavidze** [reply]
Thank you very much for replying first.
I want to run the admin dashboard and its API side separately, for now when I create an admin project, and start the server the dashboard starts with it.
The purpose is to deploy them separately.

**Reply by shahednasser** [reply]
If it's for deployment purposes, you can check out the [Vercel deployment guide in the docs](https://docs.medusajs.com/deployments/admin/deploying-on-vercel#step-2-configure-build-script).

**Reply by shavidze** [reply]
@shahednasser Thanks for your feedback, but this time, we are deploying the backend and dashboard together. Is it possible to use the [medusajs/admin-ui](https://github.com/medusajs/medusa/tree/develop/packages/admin-ui) package as a separate admin project and extend my dashboard on it and then separately deploy it and API?

**Reply by shahednasser** [reply]
I'm not sure I understand you as your requirement keeps changing.

The admin is essentially a Medusa [plugin](https://docs.medusajs.com/development/plugins/create), so you can always create a plugin with an admin dashboard if that's what you want. You can also remove the admin plugin from your `medusa-config.js` and you won't have an admin dashboard, but the admin REST APIs remain the same. So, you can build a separate dashboard that interacts with those APIs.

If you want to extend the admin, we provide extension features like [widgets](https://docs.medusajs.com/admin/quickstart), [routes](https://docs.medusajs.com/admin/routes), and [setting pages](https://docs.medusajs.com/admin/setting-pages).

If you mean something else, then please explain more.

**Reply by shavidze** [reply]
The admin dashboard is installed on the Medusa backend. The admin dashboard starts when you start the Medusa backend - as the documentation says.
I wanted to have separately this two things: __Admin Dashboard__ and __Medusa backend__, and use this __Admin Dashboard__ you are using with medusa backend as my jumpstart of my Admin Dashboard.
That is it.

**Reply by shahednasser** [reply]
I'm sorry, I'm not sure I'm following what you mean by "jumpstart of my admin dashboard". Like I mentioned, if you need to deploy it separately you can follow [this guide](https://docs.medusajs.com/deployments/admin/deploying-on-vercel#step-2-configure-build-script). You can also set some environment variable that enables/disables registering the Medusa admin in your `medusa-config.js` like so:

```js
const plugins = [
  // ...
  process.env.ENABLE_ADMIN && {
    resolve: "@medusajs/admin",
    /** @type {import('@medusajs/admin').PluginOptions} */
    options: {
      autoRebuild: true,
      develop: {
        open: process.env.OPEN_BROWSER !== "false",
      },
    },
  },
];
```

If you need additional help please try our [Discord server](https://discord.com/invite/medusajs). Otherwise, I've referred you to the docs/possible solutions based on my understanding of your use case.

**Reply by shavidze** [reply]
I just wanted to have `admin frontend` as a `standalone` project.

**Reply by JCheungX** [reply]
I have a similar case too.

There was a way to eject admin dashboard as separate project and further customize base on that.
That option is gone since the admin-ui-extension update. 
Is there any plan for that option to be available again?
we are heavily customising the dashboard , but we were hoping we can update the dashboard from the medsua team changes as well.

**Reply by shavidze** [reply]
Maybe you mean this project, I found recently https://github.com/medusajs/admin

**Reply by kasperkristensen** [reply]
As it has been mentioned we don't support ejecting the admin dashboard anymore. Ejecting the admin is not something we want to encourage, as it will result issue when trying to keep an ejected admin up to date with upstream changes to the project. Also the current admin dashboard contains logic for our extension mechanism, which built with the admin being built in the context of a Medusa server project.

Regarding your question (@JCheungX ) if we are planning on supporting the eject command again, the answer is no. We are committed to making admin extensions flexible enough that this should never be necessary. We are always interested in feedback on what extensions are currently lacking which makes you want to eject it completely, which we will then consider when planning new features for the admin and the extension mechanism.

Regarding the OPs question, if you really want to separate the admin completely from your backend, then you will need to grab the source code from here: https://github.com/medusajs/medusa/tree/develop/packages/admin-ui, and then you are free to make any changes you want. Again, we don't want to support this due to the pain points I have outlined above, so I would urge you to reconsider if you can achieve what you are looking for with admin extensions rather than ejecting completely. And also let us know what admin extensions are currently lacking which is making you want to eject the admin, we already have improvements planned, but we are always open for further suggestions and perhaps we can then also include those.

**Reply by mohammadraasel** [reply]
@kasperkristensen 
I'm looking to enhance the category creation form by adding a new field called 'thumbnail,' specifically designed for uploading image files. Could you guide me on the process of extending the form to incorporate this customization?

**Reply by crystalschang** [reply]
@kasperkristensen I've been trying to use the existing admin extensions to customize the UI, but some use cases that aren't supported with the current framework are:

* Conditionally showing/hiding the sidebar navigation in the app (e.g. based on a user's role, they should not have access to create discounts, adjust pricing, etc)
* Add additional custom fields in "Create Product" modals
* Creating new full page templates that are outside of the logged in medusa frame (e.g. a separate signup page)
* Customizing existing full page templates (e.g. add my own logo to the login/invite page)

Curious if there's any other workarounds for this and/or if there are potential plans to support in the future with Medusa 2.0?

**Reply by NicolasGorga** [reply]
@crystalschang hello, did you get any response regarding this from the team? I am seeing the same limitations since i am trying to build a marketplace with MedusaJS.

**Reply by devtony01** [reply]
@shahednasser please we need an answer. I'm trying to restrict access to certain roles on the sidebar, and minimize it also. I would also like to customize the navbar/header. Nvm. 
edit: Creating a new dashboard works just fine
re-edit: I just thought even if I create a new dashboard, how would I intergrate existing plugins?

**Reply by devtony01** [reply]
Can you give an official guide on how to do this? How to create a plugin with an admin dashboard

**Reply by shahednasser** [reply]
Creating a plugin with admin customizations is just the same as creating the customizations within the Medusa backend. For example, you create a widget in `src/admin/widgets`.

The only thing to note is that plugins having admin customizations must have the `enableUI` enabled on the backend it's installed in.

[This guide](https://docs.medusajs.com/development/plugins/create) on how to create a plugin includes the necessary details.

**Reply by devtony01** [reply]
@shahednasser thanks so much for your reply, but it doesn't answer my question of how I can create an entirely new admin dashboard whether there is an official guide for that, if not, can we customize namely: the current admin dashboard sidebar, to minimize it and restrict access based on roles and to customize the navbar/header. Thanks for your time, I know you are busy

**Reply by shahednasser** [reply]
There's no guide to creating a custom admin as this is really up to you how you do it. We provide you with the necessary tools to send requests to the backend:

- [Medusa JS Client](https://docs.medusajs.com/js-client/overview)
- [Medusa React](https://docs.medusajs.com/medusa-react/overview)
- [REST APIs](https://docs.medusajs.com/api/admin)

As for customizing the current admin dashboard, no it's currently not possible to control the sidebar other than to add new sidebar items for new pages. The same goes for the navigation bar and header.

**Reply by devtony01** [reply]
Thanks again for your prompt reply, I have another question? is medusa 2.0 ready for use? Can i also use the develop branch now?

**Reply by shahednasser** [reply]
We recommend it only for exploration/getting an idea of its features, but not for production use.

**Comment by NicolasGorga** [comment]
@kasperkristensen to add to @crystalschang answer, i am seeing those same limitations for building a marketplace and i should add:

- Setting custom logic for first navigation after login. For example, Vendors when signing up should be left in a 'pending verification' state, which sould redirect any navigation attempts to a route like '/pending-verification'.

Very much interested in your answer, since i realli liked what i've seen so far about Medusa, but find these issues as being critical to use it to build an ecommerce. Greetings from Uruguay!

**Comment by rizkyputri13** [comment]
> Is there any **separate project,** or anything related to it to run separately `admin UI` and `admin API` projects?
> 
> P.S. There was a separate project but it's archived for now https://github.com/medusajs/admin

Maybe you can see this docs...

- https://docs.medusajs.com/resources/medusa-cli/commands/build
- https://github.com/medusajs/medusa/blob/f5235862c07a6bd244a87e0f18934bbaf6390cad/packages/cli/medusa-cli/src/create-cli.ts#L422
- https://github.com/medusajs/medusa/blob/f5235862c07a6bd244a87e0f18934bbaf6390cad/packages/core/framework/src/build-tools/compiler.ts#L342
- https://github.com/medusajs/medusa/blob/develop/packages/core/framework/src/config/types.ts
